### ISSUE TITLE
KAT checks not working right on master

### ISSUE THREAD

**Issue by dstebila** [open-quantum-safe/liboqs#360]
There are several problems with the recently merged KAT checks on master:

1. Building on macOS without OpenSSL yields the following error.  This is because rand\_nist.c depends on OpenSSL, but we should be able to build liboqs master without OpenSSL.  So we need to selectively disable all relevant parts (rand\_nist, kat\_kem) when OpenSSL is not present.  Or we need to decide to require OpenSSL -- which would be something for discussion.

```
rand_nist.c:12:10: fatal error: 'openssl/conf.h' file not found
#include <openssl/conf.h>
         ^~~~~~~~~~~~~~~~
1 error generated.
```

2. Several of the Travis build logs are showing error messages despite passing.  So there are actually two problems to fix here: (a) figure out why there are KAT errors in the logs and fix them, and (b) figure out why the Travis tests aren't failing due to the errors.  See:
    - https://travis-ci.org/open-quantum-safe/liboqs/jobs/424519384
    - https://travis-ci.org/open-quantum-safe/liboqs/jobs/424519385

**Comment by dstebila** [comment]
The problem is that rand\_nist.c calls OpenSSL's AES 256 implementation via the OpenSSL EVP API.  We don't have AES 256 in our src/crypto/aes.h implementation.  

So our options are:

1. Make OpenSSL mandatory.
2. Refactor the build so rand\_nist.c and kat\_kem are not built when OpenSSL is not used.
3. Add AES 256 without OpenSSL.

3 seems like the proper choice -- and it will also allow us to build KATs on Windows.

To do 3, we need to:

- [ ] Implement AES-256 in src/crypto/aes* => Douglas
- [ ] Change rand\_nist.c's `AES256_ECB` function to call OQS AES 256 and remove OpenSSL includes etc. in this file => Shravan
- [ ] Remove `USE_OPENSSL` from around all randombytes\_nist\_kat and related stuff in src/common/rand.c => Shravan
- [ ] See if that makes it build on macOS without OpenSSL enabled => Shravan
- [ ] Add a Travis test for build on macOS without OpenSSL enabled => Shravan
- [ ] Add rand\_nist.c to VisualStudio build => Shravan, Christian
- [ ] Add kat\_kem to VisualStudio build => Christian
- [ ] Add test script to AppVeyor that checks KATs => Christian

**Comment by dstebila** [comment]
What's the status of this item?

**Comment by smashra** [comment]
Windows part needs to be done.
Other than that it should be good to go.

**Comment by christianpaquin** [comment]
I didn't have a chance to look at that yet; I'm hoping to get to it this week.

**Comment by christianpaquin** [comment]
I just took a look. rand_nist is failing because of its dependency on OpenSSL, as noted above. @smashra, do you have a PR somewhere I could look at with some refactored code pulling in @dstebila's new AES code?

I'm also wondering how important it is to build/verify KATs on Windows. The point of KATs is to make sure an alg is well implemented; and our Linux build would catch errors. The Windows build is to allow devs to use the library on Windows; ideally KAT support would be there but it's not essential, IMHO.

**Comment by smashra** [comment]
@Christian Paquin <cpaquin@microsoft.com>
Issue# 360
I just created a pull request PR#386.

I think it makes sense to have KATs on Windows.
In case someone with only Windows environment notices something we miss.
Plus it makes for a symmetrical environment.

**Comment by christianpaquin** [comment]
Perhaps we can merge PR #386 and create another issue to track adding kat_kem to Visual Studio

**Comment by smashra** [comment]
Alright then I’ll merge master to it and make it ready for merge to master.

**Comment by smashra** [comment]
Ready to merge.

**Comment by dstebila** [comment]
Resolved by #386

**Comment by dstebila** [comment]
I added issue #388 to track adding kat\_kem to VisualStudio.
